File: LEET_2398.py
from collections import deque
from typing import List


class Solution:
    def maximumRobots(self, chargeTimes: List[int], runningCosts: List[int], budget: int) -> int:
        answer = 0
        left = 0
        curr = 0
        q = deque()

        n = len(chargeTimes)
        for right in range(n):
            while q and chargeTimes[q[-1]] <= chargeTimes[right]:
                q.pop()
            q.append(right)
            curr += runningCosts[right]

            while q and chargeTimes[q[0]] + (right - left + 1) * curr > budget:
                if q[0] == left:
                    q.popleft()
                curr -= runningCosts[left]
                left += 1
            answer = max(answer, right - left + 1)

        return answer


if __name__ == "__main__":
    solution = Solution()
    chargeTimes = [[3, 6, 1, 3, 4], [11, 12, 19]]
    runningCosts = [[2, 1, 3, 4, 5], [10, 8, 7]]
    budget = [25, 19]

    for c, r, b in zip(chargeTimes, runningCosts, budget):
        print(solution.maximumRobots(c, r, b))

File: LEET_2398.cpp
// https://leetcode.com/problems/maximum-number-of-robots-within-budget/
// stacks and queues
#include <vector>
#include <algorithm>
#include <deque>
using namespace std;
class Solution {
public:
    int maximumRobots(vector<int>& chargeTimes, vector<int>& runningCosts, long long budget) {
        size_t N = chargeTimes.size();
        int left = 0;
        deque<int> dq; // indices in decreasing order
        long long cumsum = 0;
        int ans = 0;
        for (int right=0; right < N; right++) {
            // update dq 
            while (!dq.empty() && chargeTimes[dq.back()] < chargeTimes[right]){
                dq.pop_back();
            }
            dq.push_back(right);
            cumsum += runningCosts[right];
            
            // calculate weighted sum w/ current left and right
            long long cand = cumsum*(right-left+1) + chargeTimes[dq.front()];
            // resize window
            while (right >= left && cand > budget) {
                cumsum -= runningCosts[left];
                left++;
                while (!dq.empty() && dq.front() < left) {
                    dq.pop_front();
                }
                if (!dq.empty()) cand = cumsum*(right-left+1) + chargeTimes[dq.front()]; // 여기서 dq이 비는 걸 예상 못해서 에러 많이 났음!!
                //cout << "l: " << left << " r: " << right << " cumsum: " << cumsum << endl;
            }
            ans = max(ans, right-left+1);
        }
        return ans;
    }
};